### PR TITLE
add report - log

### PR DESCRIPTION
**Pull request by MackJoly**
* **What kind of changes does this PR introduce?** 
- [ ] Bug fix
- [X] new features
- [ ] performance 
- [ ] docs update

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
- [ ] Yes:
- [X] No

* **Other information**:

Report is a class which can be used as a log. At the end, Report could be export in .txt or markdown.
The aim is to use one and write what you want during a calcul for example. 
You can also add error in the report in order to stop the calcul a precise moment by using the report.error variable
Some check method can be used and filled the reporter with traceback

### CODE REVIEW

File: dessia_common/report.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        self.add_lines([line], nb_line_blank=0)
+
+    def open(self, option: str = 'a'):
+        #To avoid : Using open without explicitly specifying an encoding
+        # file = open(self.name_report + '.log', option)
+        # TODO : Mettre with open
+        if option == 'r':
+            file = open(self.name_report + '.log', 'r')
+        elif option == 'w':
+            file = open(self.name_report + '.log', 'w')
```
Comment (Tanguylo) [inline dessia_common/report.py:63]: Is this the best way to call files ?
We spoke about it: why not using a `with open` as Steven advised you ?

File: dessia_common/report.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+        return file
+
+    def close(self, file):
+        file.close()
+
+    def add_lines(self, lines: List[str], offset: int = 0, nb_line_blank: int = 0):
+        for line in range(nb_line_blank):
+            lines.append('')
```
Comment (Tanguylo) [inline dessia_common/report.py:71]: You can replace this for loop with :
`lines += [''] * nb_line_blank`

File: dessia_common/report.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        for line in lines:
+            self.content += offset*'' + line + '  \n  '
```
Comment (Tanguylo) [inline dessia_common/report.py:73]: Are you sure that `offset * ''` does something because it is strictly equal to `''` ?

I think these are equivalent:
`self.content += '  \n  '.join(lines)[6:]` 
or
`self.content += lines[0] + '  \n  '.join(lines)`

File: dessia_common/report.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+            print(line)
```
Comment (Tanguylo) [inline dessia_common/report.py:75]: Is it required to print ?

File: dessia_common/report.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+    def add_title(self, title: str):
+        self.last_offset = 0
+        title = title.upper()
+        lines = ['# ' + title]
+
+        self.add_lines(lines, nb_line_blank=2)
+
+    def add_subtitle(self, title: str):
+        self.last_offset = 0
+        title = title[0].upper() + title[1:]
```
Comment (Tanguylo) [inline dessia_common/report.py:91]: You can replace it with:
`title = title.capitalize()`

File: dessia_common/report.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+        if len(title) > self.width_line:
+            title = title[0: self.width_line]
```
Comment (Tanguylo) [inline dessia_common/report.py:93]: This is strictly equivalent to (you can remove the condition):
`title = title[0: self.width_line]`

File: dessia_common/report.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+        nb_bloc = text.split(' ')
+        lines = []
+        line = ' ' * offset
+        for bloc in nb_bloc:
+            if bloc == "\n":
+                lines.append(line)
+                line = ' ' * offset
+
+            elif len(line) + len(bloc) <= line_length:
```
Comment (Tanguylo) [inline dessia_common/report.py:120]: what happens if line_length == 2 and len(line) == 0 and bloc == "\n" ?

**Review by Tanguylo — CHANGES_REQUESTED** [review]
I think my comments seem antipathetic but they are just direct ;) So don't be afraid please :)

Globally, I understand for what purpose is this class and I am ok with such new features. 

In my comments, you will find some best practices and advises, which I guess are required for your program to be implemented in dessia_common.

File: dessia_common/report.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+        self.time_start = time.time()
+        # else:
+        #     self.time_start = time_start
+        self.last_offset = last_offset
+        self.error = False
+        self.content = content
+
+    def add_time(self, offset: int = 0, text_to_add: str = '', time_to_add: float = 0):
+        line = ' ' * offset
+        time_elapsed = str(round((time.time() - self.time_start) / 60, 2))
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:47]: Thanks for your idea, but the meaning of "add_time", is to add the time spend in a function
For example, in another script :

- import time
- init_time = time.time()
- action1
- action2
- fonction1
- report.add_time(text_to_add='function1 took', time_to_add=time.time()-init_time)

File: dessia_common/report.py [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+        if time_to_add != 0 and text_to_add != '':
+            # line += '*ELAPSED TIME min -- {} {} -- global {}*'.format(text_to_add, time_to_add, time_elapsed)
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:49]: Done

File: dessia_common/report.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+            line += f'*ELAPSED TIME min -- {text_to_add} {time_to_add} -- global {time_elapsed}*'
+        else:
+            # line += '*ELAPSED TIME min -- global {}* '.format(time_elapsed)
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:52]: Done

File: dessia_common/report.py (hunk 2 above)
Comment (MackJoly) [reply dessia_common/report.py:71]: Done

File: dessia_common/report.py (hunk 3 above)
Comment (MackJoly) [reply dessia_common/report.py:73]: Done by adding a space in the initial ''

File: dessia_common/report.py (hunk 5 above)
Comment (MackJoly) [reply dessia_common/report.py:91]: Done, thanks !

File: dessia_common/report.py (hunk 6 above)
Comment (MackJoly) [reply dessia_common/report.py:93]: Done

File: dessia_common/report.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+
+        lines = ['', '## ' + title]
+        self.add_lines(lines, nb_line_blank=1)
+
+    def add_subsubtitle(self, title: str):
+        offset = 2
+        self.last_offset = offset
+        title = title[0].upper() + title[1:]
+        if len(title) > self.width_line - offset:
+            title = title[0: self.width_line - offset]
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:103]: Done

File: dessia_common/report.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        if len(title) > self.width_line:
+            title = title[0: self.width_line]
+
+        lines = ['', '## ' + title]
+        self.add_lines(lines, nb_line_blank=1)
+
+    def add_subsubtitle(self, title: str):
+        offset = 2
+        self.last_offset = offset
+        title = title[0].upper() + title[1:]
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:101]: Done

File: dessia_common/report.py [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+        max_length = [len(t) for t in title]
+
+        for element in elements:
+            for index, elem in enumerate(element):
+                max_length[index] = max(max_length[index], len(str(elem)))
+
+        pourcent_length = [m / sum(max_length) for m in max_length]
+
+        real_length = []
+        for pourcent in pourcent_length[0:-1]:
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:160]: Done

File: dessia_common/report.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Tanguylo) [reply dessia_common/report.py:47]: `convert = time.strftime("%m months, %d days, %H:%M:%S", time.gmtime(seconds))`

is just a better way to build a time str. With it, you will know how much (months), days, hours, minutes and seconds you spent in a method and not anymore only a rounded number in minutes

File: dessia_common/report.py (hunk 6 above)
Comment (Tanguylo) [reply dessia_common/report.py]: Which is also equivalent to `title = title[: self.width_line]`, sorry

File: dessia_common/report.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Tanguylo) [reply dessia_common/report.py]: Idem, `title = title[:self.width_line - offset]`, sorry

File: dessia_common/report.py [hunk 14]
```diff
@@ -88,19 +85,19 @@ def add_title(self, title: str):
 
     def add_subtitle(self, title: str):
         self.last_offset = 0
-        title = title[0].upper() + title[1:]
-        if len(title) > self.width_line:
-            title = title[0: self.width_line]
+        title = title.capitalize()
+        # if len(title) > self.width_line:
+        title = title[0: self.width_line]
```
Comment (Tanguylo) [inline dessia_common/report.py:90]: Remove 0

File: code_pylint.py [hunk 15]
```diff
@@ -18,15 +18,15 @@
...
                      'invalid-name': 0,
                      'consider-using-f-string': 1,
                      'no-else-return': 0,
                      'arguments-differ': 2,
                      'no-member': 1,
                      'too-many-locals': 12,  # Reduce by dropping vectored objects
-                     'wrong-import-order': 7,
-                     'too-many-branches': 12,
+                     'wrong-import-order': 1,
+                     'too-many-branches': 10,
```
Comment (Tanguylo) [inline code_pylint.py:29]: Why did you change it ?
Is it a modification from a previous modification or is it to do the best as possible ?

If it is to do the best as possible, please don't modify this file, unless you have good reasons (which are that you have too many arguments or other problems like that, which can't be simply resolved except by changing the whole philosophy of your developments).

Comment (MackJoly) [reply code_pylint.py:29]: It is from master because I did a merge from master. I Needed the update from master to have cadview

File: dessia_common/report.py [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+        pourcent_length = [m / sum(max_length) for m in max_length]
+
+        real_length = []
+        for pourcent in pourcent_length[0:-1]:
+            real_length.append(int(pourcent * real_line_length))
+        real_length.append(real_line_length - sum(real_length))
+
+        lines = ['  \n  ']
+        lines.append(self.add_table_line(offset, real_length, title))
+        lines.append(self.add_table_line(offset, real_length, title, is_title = True))
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:168]: I think one method with a parameter like that is a good thing, there are less line in the code at the end
If i code two methods it would be a duplicate

File: dessia_common/report.py (hunk 4 above)
Comment (MackJoly) [reply dessia_common/report.py:71]: I added a paramater to_print in order to let the user print or not the log in the console

**Comment by MackJoly** [comment]
I move the code report.py in a new deposite called custombot on gitea. Feel free to join the deposite if you want to have an eye on the evolution. We will discuss further concerning the purpose of the deposite and how we could add it to dessia_common if it is necessary

File: dessia_common/report.py [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+
+    def add_error(self, text: str):
+        self.error = True
+        self.add_text("\n **ERROR** " + text + " \n")
+
+    def to_txt(self):
+        # TODO : connaitre le fichier comme un chemin, le réouvrir et faire ça propre
+        file = self.open('w')
+        file.write(self.content)
+        self.close(file)
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py]: Hi, yes with @younesdessia we did this method instead

File: dessia_common/report.py [hunk 18]
```diff
@@ -0,0 +1,199 @@
...
+        for element in elements:
+            for index, elem in enumerate(element):
+                max_length[index] = max(max_length[index], len(str(elem)))
+
+        pourcent_length = [m / sum(max_length) for m in max_length]
+
+        real_length = []
+        for pourcent in pourcent_length[0:-1]:
+            real_length.append(int(pourcent * real_line_length))
+        real_length.append(real_line_length - sum(real_length))
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py]: I think it is not the purpose here, the purpose is too answer " How can I manage the place left in order to add my text without cutting words"

File: dessia_common/report.py [hunk 19]
```diff
@@ -0,0 +1,199 @@
...
+        # | left foo      | right foo     |
+        # | left bar      | right bar     |
+        # | left baz      | right baz     |
+
+        offset = self.last_offset + 2
+        # line_length = self.width_line - offset
+
+        real_line_length = self.width_line - offset - (len(title) + 1) - 2 * len(title)
+        if real_line_length < 0:
+            raise KeyError('too many columns in the report')
```
[the remark this answers is not in the document]
Comment (MackJoly) [reply dessia_common/report.py:158]: Thanks for the advice, I wanted to avoid the management when the table would be bigger than the size of the line, I will think about that in a future (near or not)
